lifecycle: replaceTriggeredBy faulty

### Expected Behavior

When adding lifecycle.replaceTriggeredBy with an ITerraformDependable cdktf should generate valid terraform.

### Actual Behavior

After adding lifecycle.replaceTriggeredBy with an ITerraformDependable cdktf generates invalid terraform.

    "lifecycle": {
      "replace_triggered_by": [
        "${azurerm_subnet.network_default_subnet_6830CB96}"
      ]
    },

The terraform seems to be invalid because string interpolation seems not to be intended for that parameter.

### Steps to Reproduce

Add lifecycle.replaceTriggeredBy with an ITerraformDependable to a resource in typescript.

    lifecycle: {
      replaceTriggeredBy: [iTerraformDependable]
    }

### Versions

language: typescript
cdktf-cli: 0.19.0
node: v18.13.0
cdktf: 0.19.0
constructs: 10.3.0
jsii: null
terraform: 1.5.2
arch: x64
os: win32 10.0.22000

### Providers

┌───────────────┬──────────────────┬───────┬────────────┬──────────────┬─────────────────┐
│ Provider Name │ Provider Version │ CDKTF │ Constraint │ Package Name │ Package Version │
├───────────────┼──────────────────┼───────┼────────────┼──────────────┼─────────────────┤
│ azurerm       │ 3.76.0           │       │ ~> 3.70    │              │                 │
├───────────────┼──────────────────┼───────┼────────────┼──────────────┼─────────────────┤
│ random        │ 3.5.1            │       │ ~> 3.5     │              │                 │
├───────────────┼──────────────────┼───────┼────────────┼──────────────┼─────────────────┤
│ elastic/ec    │ 0.9.0            │       │ ~> 0.9     │              │                 │
├───────────────┼──────────────────┼───────┼────────────┼──────────────┼─────────────────┤
│ Azure/azapi   │ 1.9.0            │       │ ~> 1.9     │              │                 │
└───────────────┴──────────────────┴───────┴────────────┴──────────────┴─────────────────┘


### Gist

_No response_

### Possible Solutions

_No response_

### Workarounds

I tried the [escape hatch](https://developer.hashicorp.com/terraform/cdktf/concepts/resources#escape-hatch) but ultimately had the same issue.

### Anything Else?

_No response_

### References

_No response_

### Help Wanted

- [ ] I'm interested in contributing a fix myself

### Community Note

- Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
- Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
- If you are interested in working on this issue or have submitted a pull request, please leave a comment

I also ran into this following the docs in https://registry.terraform.io/providers/hashicorp/google/latest/docs/resources/firebaserules_release

With the `fqn` as in this issue post, it is invalid interpolation, without it, it seems to be infinite recursion.

```
[2023-11-15T17:50:29.384] [ERROR] default - RangeError: Maximum call stack size exceeded
    at keysToSnakeCase (/Users/anuraag/git/jarvis-infra/node_modules/cdktf/lib/util.ts:82:32)
    at /Users/anuraag/git/jarvis-infra/node_modules/cdktf/lib/util.ts:100:15
    at Array.reduce (<anonymous>)
    at keysToSnakeCase (/Users/anuraag/git/jarvis-infra/node_modules/cdktf/lib/util.ts:92:15)
    at /Users/anuraag/git/jarvis-infra/node_modules/cdktf/lib/util.ts:100:15
    at Array.reduce (<anonymous>)
    at keysToSnakeCase (/Users/anuraag/git/jarvis-infra/node_modules/cdktf/lib/util.ts:92:15)
    at /Users/anuraag/git/jarvis-infra/node_modules/cdktf/lib/util.ts:100:15
    at Array.reduce (<anonymous>)
```

--- a/packages/cdktf/lib/terraform-data-source.ts
+++ b/packages/cdktf/lib/terraform-data-source.ts
@@ -9,6 +9,7 @@ import {
   TerraformResourceConfig,
   TerraformResourceLifecycle,
   ITerraformResource,
+  lifecycleToTerraform,
 } from "./terraform-resource";
 import { keysToSnakeCase, deepMerge, processDynamicAttributes } from "./util";
 import { ITerraformDependable } from "./terraform-dependable";
@@ -115,7 +116,7 @@ export class TerraformDataSource
         ? this.count.toTerraform()
         : this.count,
       provider: this.provider?.fqn,
-      lifecycle: this.lifecycle,
+      lifecycle: lifecycleToTerraform(this.lifecycle),
       forEach: this.forEach?._getForEachExpression(),
     };
   }

--- a/packages/cdktf/lib/terraform-resource.ts
+++ b/packages/cdktf/lib/terraform-resource.ts
@@ -64,15 +64,17 @@ export function lifecycleToTerraform(
 
   return {
     ...lifecycle,
-    ...(lifecycle?.replaceTriggeredBy?.length && {
-      replaceTriggeredBy: lifecycle?.replaceTriggeredBy?.map((x) => {
-        if (typeof x === "string") {
-          return x;
-        } else {
-          return x.fqn;
+    ...(lifecycle?.replaceTriggeredBy?.length
+      ? {
+          replaceTriggeredBy: lifecycle?.replaceTriggeredBy?.map((x) => {
+            if (typeof x === "string") {
+              return x;
+            } else {
+              return x.fqn;
+            }
+          }),
         }
-      }),
-    }),
+      : undefined),
   };
 }
 

--- a/packages/cdktf/lib/terraform-resource.ts
+++ b/packages/cdktf/lib/terraform-resource.ts
@@ -50,6 +50,32 @@ export interface TerraformResourceLifecycle {
   readonly postcondition?: Postcondition[];
 }
 
+/**
+ * prepares a lifecycle object for being rendered as JSON
+ * currently this function:
+ *  - converts all replaceTriggeredBy items that are ITerraformDependables to strings
+ */
+export function lifecycleToTerraform(
+  lifecycle?: TerraformResourceLifecycle
+): TerraformResourceLifecycle | undefined {
+  if (!lifecycle) {
+    return undefined;
+  }
+
+  return {
+    ...lifecycle,
+    ...(lifecycle?.replaceTriggeredBy?.length && {
+      replaceTriggeredBy: lifecycle?.replaceTriggeredBy?.map((x) => {
+        if (typeof x === "string") {
+          return x;
+        } else {
+          return x.fqn;
+        }
+      }),
+    }),
+  };
+}
+
 export interface TerraformMetaArguments {
   readonly dependsOn?: ITerraformDependable[];
   readonly count?: number | TerraformCount;
@@ -190,13 +216,14 @@ export class TerraformResource
       !this.forEach || typeof this.count === "undefined",
       `forEach and count are both set, but they are mutually exclusive. You can only use either of them. Check the resource at path: ${this.node.path}`
     );
+
     return {
       dependsOn: this.dependsOn,
       count: TerraformCount.isTerraformCount(this.count)
         ? this.count.toTerraform()
         : this.count,
       provider: this.provider?.fqn,
-      lifecycle: this.lifecycle,
+      lifecycle: lifecycleToTerraform(this.lifecycle),
       forEach: this.forEach?._getForEachExpression(),
       connection: this.connection,
     };

--- a/packages/cdktf/test/resource.test.ts
+++ b/packages/cdktf/test/resource.test.ts
@@ -328,6 +328,23 @@ it("maintains the same order of provisioner", () => {
   expect(Testing.synth(stack)).toMatchSnapshot();
 });
 
+it("supports resource and attribute references in lifecycle.replaceTriggeredBy", () => {
+  const app = Testing.app();
+  const stack = new TerraformStack(app, "test");
+  new TestProvider(stack, "provider", {});
+
+  const other = new TestResource(stack, "other", { name: "other" });
+
+  new TestResource(stack, "simple", {
+    name: "foo",
+    lifecycle: {
+      replaceTriggeredBy: [other, other.stringValue],
+    },
+  });
+
+  expect(Testing.synth(stack)).toMatchSnapshot();
+});
+
 test("includes import block when import is present", () => {
   const app = Testing.app();
   const stack = new TerraformStack(app, "test");

